src/background.ts

```diff
@@ -30,24 +38,54 @@ const subscribe = (
     batching: { maxMilliseconds: 10 },
   });
   const childLogger = logger.child({ subscription });
-  sub.on('message', async (message) => {
-    try {
-      await handler(message, connection, childLogger, pubsub);
-      message.ack();
-    } catch (err) {
-      childLogger.error(
-        { messageId: message.id, data: message.data.toString('utf-8'), err },
-        'failed to process message',
-      );
-      message.nack();
-    }
+  const histogram = meter.createHistogram('message_processing_time', {
+    unit: 'ms',
+    description: 'time to process a message',
   });
+  sub.on('message', async (message) =>
+    runInRootSpan(
+      `message: ${subscription}`,
+      async (span) => {
+        const startTime = Date.now();
+        let success = true;
+        addPubsubSpanLabels(span, subscription, message);
+        try {
+          await runInSpan('handler', async () =>
+            handler(message, connection, childLogger, pubsub),
+          );
+          message.ack();
+        } catch (err) {
+          success = false;
+          childLogger.error(
+            {
+              messageId: message.id,
+              data: message.data.toString('utf-8'),
+              err,
+            },
+            'failed to process message',
+          );
+          message.nack();
+        }
+        histogram.record(Date.now() - startTime, {
```

Comment:
It's recommended to use the performance API for such use cases I believe. https://nodejs.org/api/perf_hooks.html

Reply:
I'll take a look